File: app/src/main/java/com/ravimishra/tradzhub/Adapter/ProductivityDetailAdapter.java
package com.ravimishra.tradzhub.Adapter;

import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.ravimishra.tradzhub.Model.TradzHubProductModel;
import com.ravimishra.tradzhub.R;
import com.ravimishra.tradzhub.Utils.Constants;

import java.util.List;

public class ProductivityDetailAdapter extends RecyclerView.Adapter<ProductivityDetailAdapter.viewholder> {

    List<TradzHubProductModel> productModel;
    private Context context;
    private int type = 0;

    public ProductivityDetailAdapter(Context context, List<TradzHubProductModel> productModel, int type) {
        this.context = context;
        this.productModel = productModel;
        this.type = type;
    }

    private static String removeLastChar(String str) {
        return str.substring(0, str.length() - 2);
    }

    @NonNull
    @Override
    public ProductivityDetailAdapter.viewholder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView;
        if (type == 0) {
            itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.cart_rowe_item, parent, false);
        } else {
            itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.wishlist_row, parent, false);
        }
        return new ProductivityDetailAdapter.viewholder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull viewholder holder, int position) {
        final TradzHubProductModel model = productModel.get(position);

        TradzHubProductModel.ResponseData modelData = model.data.get(0);
        RequestOptions options = new RequestOptions()
                .centerCrop()
                .placeholder(R.drawable.place_holder_image)
                .error(R.drawable.place_holder_image);
        Glide.with(context).load(model.data.get(0).productImage).apply(options).into(holder.img);

        String orignalp = modelData.salePrice;
        Float orignalPvalue = Float.parseFloat(orignalp);
        String purchaseP = modelData.purchasePrice;
        Float purchasePvalue = Float.parseFloat(purchaseP);

        int discountPrice = (int) (orignalPvalue - purchasePvalue);
        int percentDiscount = (int) ((purchasePvalue / orignalPvalue) * 100);
        int stockCount = Integer.parseInt(modelData.currentStock);
        if (stockCount != 0) {
            holder.tvAvailabilityText.setText("In Stock");
            holder.tvAvailabilityText.setTextColor(context.getResources().getColor(R.color.green));
        } else {
            holder.tvAvailabilityText.setText("Out of stock");
            holder.tvAvailabilityText.setTextColor(context.getResources().getColor(R.color.red));
        }

        if (discountPrice > 0) {
            holder.tvDiscountPrice.setText("$ " + discountPrice);
            holder.tvOffPrice.setVisibility(View.VISIBLE);
            holder.tvOffPrice.setText(percentDiscount + "% off");
        } else {
            holder.tvDiscountPrice.setText("$0");
            holder.tvOffPrice.setVisibility(View.GONE);
        }

        holder.productName.setText(modelData.title);
        holder.productPice.setText("$ " + modelData.purchasePrice);
        holder.orignalPrice.setText("$ " + modelData.salePrice);
        holder.tvOffPrice.setText(percentDiscount + "% off");
        holder.plusBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int cartCount = Integer.parseInt(holder.cartCount.getText().toString().trim());
                cartCount = cartCount + 1;
                holder.cartCount.setText(cartCount + "");
            }
        });

        holder.minusBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int cartCount = Integer.parseInt(holder.cartCount.getText().toString().trim());
                if (cartCount != 0) {
                    cartCount = cartCount - 1;
                }
                holder.cartCount.setText(cartCount + "");
            }
        });
        holder.removeImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int productId = Integer.parseInt(modelData.productID);
                String cartItem;
                SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
                cartItem = preferences.getString(Constants.INSTANCE.getSHARED_CART_ITEM(), "");
                SharedPreferences.Editor editor = preferences.edit();

                String cartString = preferences.getString(Constants.INSTANCE.getSHARED_CART_ITEM(), "");
                String updatedCart = cartString.replace(modelData.productID, "");
                String updatedCartString = updatedCart.replace(",,", ",");
                String lastItem = updatedCartString.substring(updatedCartString.length() - 1);
                if (lastItem == ",") {
                    updatedCartString = removeLastChar(updatedCartString);
                }
                editor.putString(Constants.INSTANCE.getSHARED_CART_ITEM(), updatedCartString);
                editor.apply();
                Toast.makeText(context, "Item removed", Toast.LENGTH_SHORT).show();
                ((Activity) context).finish();
                ((Activity) context).overridePendingTransition(0, 0);
                context.startActivity(((Activity) context).getIntent());
                ((Activity) context).overridePendingTransition(0, 0);
            }
        });
    }

    @Override
    public int getItemCount() {
        return productModel.size();
    }

    class viewholder extends RecyclerView.ViewHolder {

        ImageView img;
        TextView productName, productPice, orignalPrice, tvDiscountPrice, tvOffPrice, tvAvailabilityText, cartCount;
        ImageView minusBtn, plusBtn, removeImage;

        public viewholder(@NonNull View itemView) {
            super(itemView);

            img = itemView.findViewById(R.id.cartImage);
            productName = itemView.findViewById(R.id.itemName);
            productPice = itemView.findViewById(R.id.tvPrice);
            orignalPrice = itemView.findViewById(R.id.productOrignalPrice);
            tvDiscountPrice = itemView.findViewById(R.id.tvDiscountPrice);
            tvOffPrice = itemView.findViewById(R.id.tvOffPrice);
            tvAvailabilityText = itemView.findViewById(R.id.tvAvailabilityText);
            cartCount = itemView.findViewById(R.id.cartCount);
            minusBtn = itemView.findViewById(R.id.minusBtn);
            plusBtn = itemView.findViewById(R.id.plusBtn);
            removeImage = itemView.findViewById(R.id.removeItemImage);
        }
    }
}

File: app/src/main/java/com/ravimishra/tradzhub/Adapter/ProductivityDetailAdapter.kt
package com.ravimishra.tradzhub.Adapter

import android.app.Activity
import android.content.Context
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.ravimishra.tradzhub.Model.Product
import com.ravimishra.tradzhub.Model.TradzHubProductModel
import com.ravimishra.tradzhub.R
import com.ravimishra.tradzhub.Utils.Constants.SHARED_CART_ITEM

class ProductivityDetailAdapter(private val context: Context, var productModel: List<Product>, val type: Int) : RecyclerView.Adapter<ProductivityDetailAdapter.viewholder>() {
    private val inStockOptions = arrayOf(0,1,2)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): viewholder {
        val itemView: View
        if (type == 0) {
            itemView = LayoutInflater.from(parent.context).inflate(R.layout.cart_rowe_item, parent, false)
        } else {
            itemView = LayoutInflater.from(parent.context).inflate(R.layout.wishlist_row, parent, false)
        }
        return viewholder(itemView)
    }

    override fun onBindViewHolder(holder: viewholder, position: Int) {
        val model = productModel[position]
        val options = RequestOptions()
                .centerCrop()
                .placeholder(R.drawable.place_holder_image)
                .error(R.drawable.place_holder_image)
        Glide.with(context).load(model.imgUrl).apply(options).into(holder.img)
        var oriPrice = model.price.toFloat()
        var discPercent = model.discount.toFloat()
        var discountedPrc : Float = oriPrice * (1 - discPercent / 100)

        var orignalPrice = oriPrice.toInt()
        var discountPercent = discPercent.toInt()
        var discountedPrice  = discountedPrc.toInt()

//        val orignalp = modelData.salePrice
//        val orignalPvalue = orignalp.toFloat()
//        val purchaseP = modelData.purchasePrice
//        val purchasePvalue = purchaseP.toFloat()
//        val discountPrice = (orignalPvalue - purchasePvalue).toInt()
//        val percentDiscount = ((purchasePvalue / orignalPvalue) * 100).toInt()
       // val stockCount = modelData.currentStock.toInt()
        if (inStockOptions.random() != 0) {
            holder.tvAvailabilityText.text = "In Stock"
            holder.tvAvailabilityText.setTextColor(context.resources.getColor(R.color.green))
        } else {
            holder.tvAvailabilityText.text = "Out of stock"
            holder.tvAvailabilityText.setTextColor(context.resources.getColor(R.color.red))
        }
        val priceDiff = orignalPrice - discountedPrice
            holder.tvDiscountPrice.text = "$ $priceDiff"
            holder.tvOffPrice.visibility = View.VISIBLE
            holder.tvOffPrice.text = "$discountPercent% off"

        holder.productName.text = model.name
        holder.productPice.text = "₹ $discountedPrice"
        holder.orignalPrice.text = "₹ $orignalPrice"
        holder.tvOffPrice.text = "$discountPercent% off"
        holder.plusBtn.setOnClickListener(View.OnClickListener {
            var cartCount = holder.cartCount.text.toString().trim { it <= ' ' }.toInt()
            cartCount = cartCount + 1
            holder.cartCount.text = cartCount.toString() + ""
        })
        holder.minusBtn.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View) {


            }
        })
        holder.removeImage.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View) {

            }
        })
    }

    override fun getItemCount(): Int {
        return productModel.size
    }

    inner class viewholder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var img: ImageView
        var productName: TextView
        var productPice: TextView
        var orignalPrice: TextView
        var tvDiscountPrice: TextView
        var tvOffPrice: TextView
        var tvAvailabilityText: TextView
        var cartCount: TextView
        var minusBtn: ImageView
        var plusBtn: ImageView
        var removeImage: ImageView

        init {
            img = itemView.findViewById(R.id.cartImage)
            productName = itemView.findViewById(R.id.itemName)
            productPice = itemView.findViewById(R.id.tvPrice)
            orignalPrice = itemView.findViewById(R.id.productOrignalPrice)
            tvDiscountPrice = itemView.findViewById(R.id.tvDiscountPrice)
            tvOffPrice = itemView.findViewById(R.id.tvOffPrice)
            tvAvailabilityText = itemView.findViewById(R.id.tvAvailabilityText)
            cartCount = itemView.findViewById(R.id.cartCount)
            minusBtn = itemView.findViewById(R.id.minusBtn)
            plusBtn = itemView.findViewById(R.id.plusBtn)
            removeImage = itemView.findViewById(R.id.removeItemImage)
        }
    }

    companion object {
        private fun removeLastChar(str: String): String {
            return str.substring(0, str.length - 2)
        }
    }


}